closedts: delete old closedts mechanism

andreimatei opened the pull request:
Deletion of old code followed by cleanups. See individual commits.
The last commit provides a significant improvement by permitting follower reads under an invalid lease.

Release justification: Too good to wait another release. Also the last commit is needed for bounded staleness read availability
during partial failures.

andreimatei committed:
closedts: delete old closedts mechanism

Closes #61989
Closes #60682
Closes #59448

andreimatei committed:
kvserver: prepare a test for the future

This test wanted to know a request's LAI at evaluation time. A future
patch is moving LAI assignment to proposal buffer flush time. Luckily,
this test didn't have a good reason for doing what it wanted.

Release note: None

andreimatei committed:
kvserver: move LAI assignment to flush time

We used to assign LAIs on proposals when the proposals were inserted
into the proposal buffer. Knowing the LAI at evaluation time was needed
by the old closed timestamp mechanism. Now that that mechanism is gone,
the assignment can move to the proposal buffer flush. Moving it there
mirrors the assignment of the closed timestamp, which is in relationship
with the LAI. This move also lets us simplify the proposal buffer code
significantly since atomically assigning an index into the propBuf's
array and an LAI is no longer needed. These assignments were done in a
semi-lock-free manner, and so doing it atomically was complex.

Release note: None

andreimatei committed:
kvserver: minor test API change

Switch the API of the test Raft group, allowing it to accumulate
proposals from multiple proposal buffer flushes. This is anticipation of
more use of this mock.

Release note: None

andreimatei committed:
kvserver: serve follower reads when lease is invalid

Before this patch, requests were not eligible for follower reads when
the lease was invalid (or when the follower replica in question thought
that there's no valid lease because it's behind). The reasons for this
bizarre state of affairs stem historically from the intertwining of
leases and closed timestamps - the current leaseholder had to be known
in order to ask it for a closed timestamp. This entanglement has now
gone away, so we can serve follower reads regardless of the lease
status as nature intended.

Closes #57992

Release note: None
Release justification: needed for bounded staleness read availability
during partial failures.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69121)
<!-- Reviewable:end -->

andreimatei committed:
kvclient: make a test more robust

TestFollowerReadsWithStaleDescriptor was fragile: it wants to control
precisely which follower a particular query will be routed to and it
went through some knobs to do so, but one more knob is needed. This
patch introduces the knob for inhibiting the GRPCTransport's check of
connection health when deciding the replica ordering - it normally tries
to deprioritize nodes to which there isn't a healthy conn. A connection
starts off as unhealthy until it's heartbeated for the first time so,
depending on whether a connection was established or not to the desired
node sufficiently before the query of interest, that check is a problem.

In particular, the transport's default behavior becomes a problem with
the following commits which remove the old closed timestamp mechanism
that came with early establishment of some network connections. Without
it, this test becomes unhappy.

Release note: None

nvb reviewed (commented):
> Reviewed 4 of 4 files at r1, 80 of 80 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 11 of 11 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69121)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 292 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht-O4xCAaL8200sae5:-Mht-O4xCAaL8200sae6:bdqyv0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f092ad9cc4f032f26bd02ff4c7e209cccfe46fde/pkg/kv/kvclient/kvcoord/dist_sender.go#L292)):*
> ```Go
> 	// only applies to requests sent with the LEASEHOLDER routing policy.
> 	dontReorderReplicas bool
> 	// dontConsiderConnectionHealth, if set, makes the GRPCTransport not take into
> ```

`dontConsiderConnHealth`
___
*[pkg/kv/kvclient/kvcoord/transport.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht-X5NAqcbaKX80Pts:-Mht-X5NAqcbaKX80Ptt:b-zax976) ([raw file](https://github.com/cockroachdb/cockroach/blob/f092ad9cc4f032f26bd02ff4c7e209cccfe46fde/pkg/kv/kvclient/kvcoord/transport.go#L37)):*
> ```Go
> 	class   rpc.ConnectionClass
> 	metrics *DistSenderMetrics
> 	// dontConsiderConnectionHealth, if set, makes the transport not take into
> ```

nit: might as well keep this as `dontConsiderConnHealth` as well, to allow for easier greping.
___
*[pkg/kv/kvserver/replica.go, line 1174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht0-_m-ndnVRU6PWuW:-Mht0-_m-ndnVRU6PWuX:b-mv95ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica.go#L1174)):*
> ```Go
> 	ri.RangeMaxBytes = *r.mu.zone.RangeMaxBytes
> 	if desc := ri.ReplicaState.Desc; desc != nil {
> 		if r.mu.tenantID != (roachpb.TenantID{}) {
> ```

Does this need to be in the `desc != nil {` branch?
___
*[pkg/kv/kvserver/replica_closedts_internal_test.go, line 569 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht0hvUE_j1b7oyz0QM:-Mht0hvUE_j1b7oyz0QN:b-cvvsp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica_closedts_internal_test.go#L569)):*
> ```Go
> 			tc.repl.mu.state.LeaseAppliedIndex = uint64(test.applied)
> 			tc.repl.mu.Unlock()
> 			require.Equal(t, test.expClosed, tc.repl.maxClosed(ctx))
> ```

Should we call `GetClosedTimestamp`? That's what the test says it tests.
___
*[pkg/kv/kvserver/replica_follower_read.go, line 122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht1_1_9q0gMCJjHsyI:-Mht1_1_9q0gMCJjHsyJ:b-96tzdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica_follower_read.go#L122)):*
> ```Go
> // cache forward to at least the new lease start time. Using this combination
> // allows the closed timestamp mechanism to be robust to lease transfers.
> func (r *Replica) maxClosed(ctx context.Context) hlc.Timestamp {
> ```

Any reason to keep this around instead of folding it into `GetClosedTimestamp`?
___
*[pkg/kv/kvserver/replica_follower_read.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht1ANe7v7Gy1-C8K7Q:-Mht1ANe7v7Gy1-C8K7R:bfyzvmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica_follower_read.go#L137)):*
> ```Go
> 	appliedLAI := ctpb.LAI(r.mu.state.LeaseAppliedIndex)
> 	leaseholder := r.mu.state.Lease.Replica.NodeID
> 	initialMaxClosed := r.mu.initialMaxClosed
> ```

Can `r.mu.initialMaxClosed` be deleted now as well?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtNqfUE9hDsFn595UB:-MhtNqfUE9hDsFn595UC:bmqi51t) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L68)):*
> ```Go
> 	// arr contains the buffered proposals.
> 	arr propBufArray
> 	// allocatedIdx is the last allocated index into propBufArray. Accessed
> ```

It's the next index to allocate, right?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht4Ysq29oq4ico6IBE:-Mht4Ysq29oq4ico6IBF:b-85m2b8) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L92)):*
> ```Go
> 	// Buffers used to avoid allocations.
> 	tmpLAIFooter             kvserverpb.MaxLeaseFooter
> 	tmpClosedTimestampFooter kvserverpb.ClosedTimestampFooter
> ```

Should we combine these? Any reason to keep them separated? Is the idea that we don't want to encode a zero-valued closed timestamp when one is not meant to be set at all? Or is it just convenient to marshal twice? If possible, it would be nice to combine these into a single marshal in the vastly common case when both fields are set. So if we need to, maybe have a `tmpLAIAndClosedTimestampFooter` and a `tmlLAIFooter`?

If you make this change, I suspect that you'll want to split up `assignClosedTimestampAndLAIToProposalLocked` into a method that allocates and returns the lai and closed timestamp, and then one that marshalls the returned lai and (optional) closed timestamp.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht4gUx22FN_A0XAhGd:-Mht4gUx22FN_A0XAhGe:bahe2rf) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L92)):*
> ```Go
> 	// Buffers used to avoid allocations.
> 	tmpLAIFooter             kvserverpb.MaxLeaseFooter
> 	tmpClosedTimestampFooter kvserverpb.ClosedTimestampFooter
> ```

nit: let's keep the relative order between LAI and closed timestamp fixed between fields/params/method names/etc. I think it makes sense to order the `LAI` first, since it's assigned first.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 194 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht5AzdAwsD9M93wG4D:-Mht5AzdAwsD9M93wG4E:bf2t9xn) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L194)):*
> ```Go
> }
> 
> func (b *propBuf) incAllocatedIdx() int64 {
> ```

nit: want to just return an `int` here, since you're doing that from `AllocatedIdx`?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 194 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtNeh9C2LHQ2CsmJLC:-MhtNeh9C2LHQ2CsmJLD:b-5o2zc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L194)):*
> ```Go
> }
> 
> func (b *propBuf) incAllocatedIdx() int64 {
> ```

Could you give this a comment?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 194 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtNhRi6w1wC-k2BLSS:-MhtNhRi6w1wC-k2BLST:b-iw1uco) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L194)):*
> ```Go
> }
> 
> func (b *propBuf) incAllocatedIdx() int64 {
> ```

Want to return -1 from this method so that we don't need to do so below? When you add a comment, I suspect you'll find this easier to explain.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 365 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtJunq8P86KRoIf9Gi:-MhtJunq8P86KRoIf9Gj:b-d9vi15) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L365)):*
> ```Go
> 	// consumers.
> 	used := b.AllocatedIdx()
> 	b.clearAllocatedIdx()
> ```

Want to have `clearAllocatedIdx` return the number of elements used like it did before, using `SwapInt64`? That avoids an extra atomic operation, and more importantly, doesn't raise red flags because it doesn't look racy (even though this isn't due to the locking).
___
*[pkg/kv/kvserver/replica_raft.go, line 90 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtLHzI0hRk6IGDm2uq:-MhtLHzI0hRk6IGDm2ur:b-ea7y5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_raft.go#L90)):*
> ```Go
> }
> 
> // evalAndProposeInner is like evalAndPropose, except it also returns the
> ```

I'm slightly :thumbsdown: on this. It doesn't seem like a good use of an additional function call on a hot path when we can just return the `kvserverbase.CmdIDKey` from `evalAndPropose` and ignore it if we want to.
___
*[pkg/kv/kvserver/replica_send.go, line 585 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtNB-z4Gh3zncZNPKW:-MhtNB-z4Gh3zncZNPKX:b3wkuhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f2c4ad795f0c9922172fcbed5932ef3b05fa625/pkg/kv/kvserver/replica_send.go#L585)):*
> ```Go
> 	// redirect.
> 	_, pErr := r.redirectOnOrAcquireLeaseForRequest(ctx, ba.Timestamp)
> 	// If we managed to get a lease, the request evaluation will be retried.
> ```

```
If we managed to get a lease (pErr == nil), ...
```
___
*[pkg/kv/kvserver/replica_write.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht2GdDCZRYAQNeRYY5:-Mht2GdDCZRYAQNeRYY6:b-i5c7cx) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica_write.go#L104)):*
> ```Go
> 	// starts being tracked after we apply the timestamp cache.
> 	var tok TrackedRequestToken
> 	var minTS hlc.Timestamp
> ```

nit: re-order these to match the return value order.
___
*[pkg/kv/kvserver/closedts/ctpb/service.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht35j37q71d9qBQWuk:-Mht35j37q71d9qBQWul:b-xbkwrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/closedts/ctpb/service.go#L24)):*
> ```Go
> 
> // SeqNum identifies a ctpb.Update.
> type SeqNum int64
> ```

Should this implement `redact.SafeValue` as well?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69121)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 292 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht-O4xCAaL8200sae5:-Mhz7yjH9BODXj1Zq8ZB:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f092ad9cc4f032f26bd02ff4c7e209cccfe46fde/pkg/kv/kvclient/kvcoord/dist_sender.go#L292)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`dontConsiderConnHealth`
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/transport.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht-X5NAqcbaKX80Pts:-Mhz7wHrAZ7kTq6Kk5mT:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f092ad9cc4f032f26bd02ff4c7e209cccfe46fde/pkg/kv/kvclient/kvcoord/transport.go#L37)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: might as well keep this as `dontConsiderConnHealth` as well, to allow for easier greping.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica.go, line 1174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht0-_m-ndnVRU6PWuW:-MhzALgI717Z08FLNmr8:b4i83g4) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica.go#L1174)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this need to be in the `desc != nil {` branch?
</blockquote></details>

I don't really know. I bet it doesn't. Fixing separately in https://github.com/cockroachdb/cockroach/pull/69389
___
*[pkg/kv/kvserver/replica_closedts_internal_test.go, line 569 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht0hvUE_j1b7oyz0QM:-MhzAh2U9P9rcBdINdzL:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica_closedts_internal_test.go#L569)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we call `GetClosedTimestamp`? That's what the test says it tests.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_follower_read.go, line 122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht1_1_9q0gMCJjHsyI:-MhzErwU7p-GDa3JFumx:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica_follower_read.go#L122)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Any reason to keep this around instead of folding it into `GetClosedTimestamp`?
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_follower_read.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht1ANe7v7Gy1-C8K7Q:-MhzNnNE7bq85tr0ELOn:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica_follower_read.go#L137)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can `r.mu.initialMaxClosed` be deleted now as well?
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtNqfUE9hDsFn595UB:-MhzR3MeAzKwCC8CaCwu:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L68)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's the next index to allocate, right?
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht4gUx22FN_A0XAhGd:-MhzRDe8CmnuSNEkxHf5:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L92)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: let's keep the relative order between LAI and closed timestamp fixed between fields/params/method names/etc. I think it makes sense to order the `LAI` first, since it's assigned first.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht4Ysq29oq4ico6IBE:-Mi30-G_8sYzO94ZtKi8:bxyv056) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L92)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we combine these? Any reason to keep them separated? Is the idea that we don't want to encode a zero-valued closed timestamp when one is not meant to be set at all? Or is it just convenient to marshal twice? If possible, it would be nice to combine these into a single marshal in the vastly common case when both fields are set. So if we need to, maybe have a `tmpLAIAndClosedTimestampFooter` and a `tmlLAIFooter`?

If you make this change, I suspect that you'll want to split up `assignClosedTimestampAndLAIToProposalLocked` into a method that allocates and returns the lai and closed timestamp, and then one that marshalls the returned lai and (optional) closed timestamp.
</blockquote></details>

done

I went back and forth on the nullability of the ClosedTimestamp field in the footer. Please look at it.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 194 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht5AzdAwsD9M93wG4D:-MhzQYyR4m9Byz1DdfE6:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L194)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: want to just return an `int` here, since you're doing that from `AllocatedIdx`?
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 194 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtNeh9C2LHQ2CsmJLC:-MhzRF3c6j9kvSD_U38D:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L194)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you give this a comment?
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 194 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtNhRi6w1wC-k2BLSS:-MhzRFmkEZcNCW2bK5QV:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L194)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Want to return -1 from this method so that we don't need to do so below? When you add a comment, I suspect you'll find this easier to explain.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 365 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtJunq8P86KRoIf9Gi:-MhzRlP225HSHtUgsU21:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L365)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Want to have `clearAllocatedIdx` return the number of elements used like it did before, using `SwapInt64`? That avoids an extra atomic operation, and more importantly, doesn't raise red flags because it doesn't look racy (even though this isn't due to the locking).
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_raft.go, line 90 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtLHzI0hRk6IGDm2uq:-MhzSfS16-TXB5nqS4eR:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_raft.go#L90)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm slightly :thumbsdown: on this. It doesn't seem like a good use of an additional function call on a hot path when we can just return the `kvserverbase.CmdIDKey` from `evalAndPropose` and ignore it if we want to.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_send.go, line 585 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MhtNB-z4Gh3zncZNPKW:-Mi30SlF3ofGLw6hZI6F:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f2c4ad795f0c9922172fcbed5932ef3b05fa625/pkg/kv/kvserver/replica_send.go#L585)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

```
If we managed to get a lease (pErr == nil), ...
```

</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_write.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht2GdDCZRYAQNeRYY5:-MhzONqC3NyvPni4aLOf:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/replica_write.go#L104)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: re-order these to match the return value order.
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/ctpb/service.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht35j37q71d9qBQWuk:-MhzOGVY97AjGt50x8iC:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/15636bc42d2af87b8a55d54f738a435a09745991/pkg/kv/kvserver/closedts/ctpb/service.go#L24)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this implement `redact.SafeValue` as well?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 3 of 91 files at r7, 71 of 80 files at r8, 1 of 2 files at r9, 14 of 14 files at r11, 5 of 5 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69121)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht4Ysq29oq4ico6IBE:-Mi3Y2I78EFJX490gSGw:b2wo1a1) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L92)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

done

I went back and forth on the nullability of the ClosedTimestamp field in the footer. Please look at it.
</blockquote></details>

I think it looks good, though I haven't audited the Raft application code to make sure it treats nil ClosedTimestamps and zero ClosedTimestamps identically. Did you look into that?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 570 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3YVv35in2_Xzl_4lB:-Mi3YVv35in2_Xzl_4lC:br7jz1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/replica_proposal_buf.go#L570)):*
> ```Go
> }
> 
> // marshallClosedTimestampAndLAIToProposalLocked modifies p.encodedCommand,
> ```

A few nits here that you're not going to like:
1. assignClosedTimestampAndLAILocked is called first, so move it above marshallClosedTimestampAndLAIToProposalLocked
2. the argument/return value order to each of these methods is `..., lai, closedTimestamp, ...`, so reflect that in their names
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 578 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3ZAdBEpIDzkIA0myR:-Mi3ZAdBEpIDzkIA0myS:b-inzjju) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/replica_proposal_buf.go#L578)):*
> ```Go
> 	// Reset the buffer, as a safety measure.
> 	*buf = kvserverpb.RaftCommandFooter{}
> 	preLen := len(p.encodedCommand)
> ```

nit: pull this right above `p.encodedCommand` so that it's clearer why it's captured.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 586 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3Z0C19g9j5fBvmK9v:-Mi3Z0C19g9j5fBvmK9w:be7ogvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/replica_proposal_buf.go#L586)):*
> ```Go
> 
> 	// Fill in the closed ts in the proposal.
> 	if !closedTimestamp.IsEmpty() {
> ```

Why the empty check? Isn't this identical?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 632 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3ZgdxECfDrJV339Q3:-Mi3ZgdxECfDrJV339Q4:b6k3h8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/replica_proposal_buf.go#L632)):*
> ```Go
> 	if filter := b.testing.leaseIndexFilter; filter != nil {
> 		if override := filter(p); override != 0 {
> 			b.assignedLAI = override
> ```

Do we want this to actually mutate `b.assignedLAI` so that it affects future proposals as well? We may, I'm just not sure and the comment on `leaseIndexFilter` doesn't mention this effect.
___
*[pkg/kv/kvserver/store_split.go, line 32 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3X-Oi4rnkMO0IzMBM:-Mi3X-Oi4rnkMO0IzMBN:bhxtoy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/store_split.go#L32)):*
> ```Go
> // split commit.
> //
> // initClosedTS is the closed timestamp carried
> ```

Carried by what?
___
*[pkg/kv/kvserver/store_split.go, line 123 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3X74J1n6Ih5ROsl_j:-Mi3X74J1n6Ih5ROsl_k:b-ye4fzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/store_split.go#L123)):*
> ```Go
> 	//
> 	// In order to tolerate a nil initClosedTS input, let's forward to
> 	// r.GetClosedTimestamp. Generally, initClosedTS is not expected to be nil
> ```

Make a note that GetClosedTimestamp itself does not account for the closed timestamp carried on the split raft command, as that command has not been applied yet.
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 323 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3XqPi4xzArGgAJG93:-Mi3XqPi4xzArGgAJG94:b-xcq4vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L323)):*
> ```protobuf
> }
> 
> // MaxLeaseFooter contains a subset of the fields in RaftCommand. It is used
> ```

This comment needs an update.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
It turns out that the `serve follower reads when lease is invalid` is important for bounded staleness read availability during partial failures. I'm looking forward to that landing.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69121)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mht4Ysq29oq4ico6IBE:-MiNG1zR-IzL5BjMx_Tl:bio40vn) ([raw file](https://github.com/cockroachdb/cockroach/blob/67fdbc82a686cfbd2160ac52166d21634d7b7a80/pkg/kv/kvserver/replica_proposal_buf.go#L92)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think it looks good, though I haven't audited the Raft application code to make sure it treats nil ClosedTimestamps and zero ClosedTimestamps identically. Did you look into that?
</blockquote></details>

I did, and it looks fine.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 570 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3YVv35in2_Xzl_4lB:-MiMjV2M0rXKHEpI84Ss:b-gq6lss) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/replica_proposal_buf.go#L570)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

A few nits here that you're not going to like:
1. assignClosedTimestampAndLAILocked is called first, so move it above marshallClosedTimestampAndLAIToProposalLocked
2. the argument/return value order to each of these methods is `..., lai, closedTimestamp, ...`, so reflect that in their names
</blockquote></details>

done and done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 578 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3ZAdBEpIDzkIA0myR:-MiMjgLHCEVrnZhlzA-u:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/replica_proposal_buf.go#L578)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: pull this right above `p.encodedCommand` so that it's clearer why it's captured.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 586 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3Z0C19g9j5fBvmK9v:-MiMkrYg-Tgzs6twbMaF:b-yey5fw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/replica_proposal_buf.go#L586)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why the empty check? Isn't this identical?
</blockquote></details>

removed the check, and removed the `*buf = kvserverpb.RaftCommandFooter{}` above too.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 632 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3ZgdxECfDrJV339Q3:-MiMmk2eAxztABW7AUEi:b-tr1g5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/replica_proposal_buf.go#L632)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we want this to actually mutate `b.assignedLAI` so that it affects future proposals as well? We may, I'm just not sure and the comment on `leaseIndexFilter` doesn't mention this effect.
</blockquote></details>

I'm not sure either (I'm not sure anyone does know)... I've changed this code to not mutate it, to keep it like it was.
___
*[pkg/kv/kvserver/store_split.go, line 32 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3X-Oi4rnkMO0IzMBM:-MiN72c1DPSYsIFat89q:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/store_split.go#L32)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Carried by what?
</blockquote></details>

done
___
*[pkg/kv/kvserver/store_split.go, line 123 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3X74J1n6Ih5ROsl_j:-MiN7S9Z5ChMaKaq53oe:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/store_split.go#L123)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Make a note that GetClosedTimestamp itself does not account for the closed timestamp carried on the split raft command, as that command has not been applied yet.
</blockquote></details>

done
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 323 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-Mi3XqPi4xzArGgAJG93:-MiNFaooCYgOHN8LAzag:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2242da901dc664e38ad4ae196feeb2b9650e788a/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L323)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This comment needs an update.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 71 of 80 files at r14, 1 of 2 files at r15, 14 of 14 files at r17, 5 of 5 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-:-MiNleK67_Bvm9v7XKbw:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 698 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MiNlDIQ6jQ1QIN66oIA:-MiNlDIQ6jQ1QIN66oIB:baa40c3) ([raw file](https://github.com/cockroachdb/cockroach/blob/eda010fd82c7d8b7c720a9dde1f6df8cd9a24d7c/pkg/kv/kvserver/replica_proposal_buf.go#L698)):*
> ```Go
> 	buf.ClosedTimestamp = closedTimestamp
> 
> 	footerLen := buf.Size()
> ```

nit: consider inlining this.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 708 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MiNl91u22ZtOF704fIj:-MiNl91u22ZtOF704fIk:bho3ikd) ([raw file](https://github.com/cockroachdb/cockroach/blob/eda010fd82c7d8b7c720a9dde1f6df8cd9a24d7c/pkg/kv/kvserver/replica_proposal_buf.go#L708)):*
> ```Go
> 	preLen := len(p.encodedCommand)
> 	p.encodedCommand = p.encodedCommand[:preLen+footerLen]
> 	// NOTE(andrei): We don't assigned to the in-memory command
> ```

nit: move this comment to right after the assignment to `buf.ClosedTimestamp`.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-:-MiObLY65xpcFW4LV0Sg:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 698 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MiNlDIQ6jQ1QIN66oIA:-MiOPOof26-ooT9e5p2G:b-od5eh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/eda010fd82c7d8b7c720a9dde1f6df8cd9a24d7c/pkg/kv/kvserver/replica_proposal_buf.go#L698)):*
> ```
> footerLen := buf.Size()
> ```
done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 708 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/69121#-MiNl91u22ZtOF704fIj:-MiOPVq34CZ0gn0sqaFs:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/eda010fd82c7d8b7c720a9dde1f6df8cd9a24d7c/pkg/kv/kvserver/replica_proposal_buf.go#L708)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move this comment to right after the assignment to `buf.ClosedTimestamp`.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->
